airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/ShutdownUtils.kt

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.core.util
+
+import com.google.common.annotations.VisibleForTesting
+import io.airbyte.commons.string.Strings
+import io.github.oshai.kotlinlogging.KotlinLogging
+import jakarta.inject.Singleton
+import org.apache.commons.lang3.ThreadUtils
+import org.apache.commons.lang3.concurrent.BasicThreadFactory
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.function.Predicate
+import java.util.stream.Collectors
+import kotlin.system.exitProcess
+
+private val logger = KotlinLogging.logger {}
+
+/**
+ * Collection of utility methods to ensure that any threads launched by a connector are cleaned up when an
+ * operation completes.
+ */
+@Singleton
+class ShutdownUtils {
+    companion object {
+        private const val TYPE_AND_DEDUPE_THREAD_NAME: String = "type-and-dedupe"
+        private const val FORCED_EXIT_CODE: Int = 2
+        const val EXIT_THREAD_DELAY_MINUTES: Int = 70
+        val EXIT_HOOK: Runnable = Runnable { exitProcess(FORCED_EXIT_CODE) }
+        const val INTERRUPT_THREAD_DELAY_MINUTES: Int = 60
+
+        /**
+         * Filters threads that should not be considered when looking for orphaned threads at shutdown of
+         * the integration runner.
+         *
+         *
+         *
+         * **N.B.** Daemon threads don't block the JVM if the main `currentThread` exits, so they are not
+         * problematic. Additionally, ignore database connection pool threads, which stay active so long as
+         * the database connection pool is open.
+         */
+        @VisibleForTesting
+        val ORPHANED_THREAD_FILTER: Predicate<Thread> =
+            Predicate<Thread> { runningThread: Thread ->
+                (
+                    runningThread.name != Thread.currentThread().name &&
+                        !runningThread.isDaemon && TYPE_AND_DEDUPE_THREAD_NAME != runningThread.name
+                )
+            }
```

Comment:
I'm worried that this approach isn't going to age well. Can we instead maintain a list of thread names that _shouldn't_ be shut down and nuke anything not matching the list?